### PR TITLE
Load paths hash

### PR DESCRIPTION
**Pull request by mclark**
## What are you trying to accomplish?

As described in https://github.com/Shopify/packwerk/issues/186, Zeitwerk supports the notion of a "default" namespace for each root directory. This is represented by a `Hash` of root directory path strings to constants representing the default module in `Zeitwerk::Loader#root_dirs`. This pull request updates Packwerk so that it treats the group of load paths as a hash rather than an Array and uses the `#root_dirs` of each autoloader in `Rails.autoloaders` instead.

## What approach did you choose and why?

As Packwerk relies on Rails' `config.autoload_paths` and friends lists to get the list of load paths to draw on for its constant discovery process, and these are just lists of strings, they needed to be abandoned in favour of relying on the autoloaders instead which are hashes of root dir paths to default modules. This triggered a number of changes, mostly to method signatures as the new hash worked its way through the call stacks to eventually arrive in the `ConstantResolver` gem (already updated in https://github.com/Shopify/constant_resolver/pull/27).

After that, most of the effort just went into updating the tests to conform to the new expectations:

* Removed `RailsPaths` in favour of using plain old Mocha mocks
* Removed uniqueness and method signature tests that were impossible to fail thanks to the use of Hashes and Sorbet
* Initialized the dummy application so `Rails.autoloaders` would be populated
* Changed two tests from `ActiveSupport::TestCase` to `Minitest::Test` to be consistent with all other tests and to suppress log output caused by initializing the dummy application.

## What should reviewers focus on?

I believe @exterm is bumping `constant_resolver` as I write this so expect the git dependency for `constant_resolver` to go away and be replaced once that is done.

I am not an expert on rails or load paths. Is eager loading still supported properly? Can we suppress the logging in a better way than outputting "nothing" to standard out?

## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Non-breaking change (a change that doesn't alter functionality - i.e., code refactor, configs, etc.)

### Additional Release Notes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)

Include any notes here to include in the release description. For example, if you selected "breaking change" above, leave notes on how users can transition to this version.

If no additional notes are necessary, delete this section or leave it unchanged.

## Checklist

- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] It is safe to rollback this change.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -9,7 +9,8 @@ gemspec
 
 gem("spring")
 gem("rails", github: "rails/rails", branch: "main")
-gem("constant_resolver", require: false)
+gem("constant_resolver", git: "https://github.com/Shopify/constant_resolver.git",
+    ref: "ba7fd11288566e2ecdd70e0946dcfb20c5341b27", require: false)
```
Comment (exterm) [inline Gemfile:13]: 0.2.0 is released now.

Comment (mclark) [reply Gemfile]: Updated! Thank you! ❤️

File: test/support/rails_application_fixture_helper.rb [hunk 2]
```diff
@@ -28,36 +18,28 @@ def use_template(template)
...
+    Rails.application.config.stubs(:root).returns(root)
   end
 
   private
 
   def set_load_paths_for_minimal_template
-    RailsPaths.autoload(to_app_paths("/components/sales/app/models"))
-    RailsPaths.eager_load([])
-    RailsPaths.autoload_once([])
+    main = Zeitwerk::Loader.new
```
Comment (mclark) [inline test/support/rails_application_fixture_helper.rb:28]: I overlooked the fact that this is introducing a direct dependency on Zeitwerk here until I saw [these tests failing](https://github.com/Shopify/packwerk/runs/5792482153?check_suite_focus=true). Given that Packwerk depends on Zeitwerk's structural assumptions (and indeed is even named after it?) I don't think adding an explicit dependency on Zeitwerk would be that much of a stretch. However, I'm tempted to just use a mock object since we only need the one `:root_dirs` method here anyway.

Do others have thoughts on the best direction here?

File: lib/packwerk/application_load_paths.rb [hunk 3]
```diff
@@ -18,36 +18,30 @@ def extract_relevant_paths(root, environment)
...
-        Rails.application.railties
-          .select { |railtie| railtie.is_a?(Rails::Engine) }
-          .push(Rails.application)
-          .flat_map do |engine|
-            paths = (engine.config.autoload_paths + engine.config.eager_load_paths + engine.config.autoload_once_paths)
-            paths.map(&:to_s).uniq
-          end
+        Rails.autoloaders.inject({}) do |h, loader|
+          h.merge(loader.root_dirs)
+        end
```
[the remark this answers is not in the document]
Comment (mclark) [reply lib/packwerk/application_load_paths.rb:25]: So it turns out this doesn't work in Rails 6.0 unless `Rails.autoloaders.zeitwerk_enabled?` returns `true`. If it is false, `Rails.autoloaders` will be an empty enumerable.

So now we have to choose one of:

* For all the places that we stub out `config.autoload_paths` and friends we add matching tests to cover when zeitwerk is both enabled and disabled.
* Explicitly add the requirement that packwerk is only compatible with Rails applications which are zeitwerk enabled. This is the case for Rails 6.x codebases that have `Rails.autoloaders.zeitwerk_enabled? == true` or Rails 7.0+ codebases.

Personally, based on our discussions about adding Zeitwerk as a dev dependency in the [test in this diff](https://github.com/Shopify/packwerk/pull/190/files#r840817911) I feel we could pretty safely say that Packwerk requires `zeitwerk_enabled?` to be true. This would retain Rails 6.x support without requiring us to duplicate a whole bunch of test.

As ["Classic mode"](https://github.com/rails/rails/commit/0d523d83657ce7066f25d87f6f094e804590e1e8) isn't coming back, and Packwerk is based on the conventions of Zeitwerk, unless this causes problems for any known users, I'd suggest we go with the addition of this "soft" requirement.

File: test/rails_test_helper.rb [hunk 4]
```diff
@@ -16,4 +18,7 @@ def self.skeleton(*path)
     skeleton("vendor", "cache", "gems", "example", "models"),
   ]
   config.root = skeleton(".")
+  config.logger = Logger.new(STDOUT)
 end
+
+T.cast(Dummy, Rails::Application).initialize!
```
Comment (exterm) [inline test/rails_test_helper.rb:24]: This assumes `Dummy` to be a value of type `Rails::Application`, but it is not - it is a subclass.

**Comment by mclark** [comment]
> Looks like we don't have a type spec for Rails::Application... maybe tapioca gem rails can fix it?

I couldn't really get it working, probably my lack of experience with `tapioca`.. Anyway, using `Rails.application.initialize!` instead seems to have done the trick 👍

File: lib/packwerk/application_load_paths.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (exterm) [reply lib/packwerk/application_load_paths.rb:25]: Maybe we should make it more obvious, but packwerk already officially requires zeitwerk to be enabled https://github.com/Shopify/packwerk#prerequisites

I'll accept a PR to reword that to make it more obvious

File: test/rails_test_helper.rb [hunk 5]
```diff
@@ -8,6 +8,8 @@ def self.skeleton(*path)
     ROOT.join("test", "fixtures", "skeleton", *path).to_s
   end
 
+  config.autoloader = :zeitwerk
```
Comment (exterm) [inline test/rails_test_helper.rb:11]: 👍🏼 This is the correct solution IMO

File: test/const_node_inspector_test.rb [hunk 6]
```diff
@@ -5,7 +5,7 @@
 require "parser_test_helper"
 
 module Packwerk
-  class ConstNodeInspectorTest < ActiveSupport::TestCase
+  class ConstNodeInspectorTest < Minitest::Test
```
Comment (exterm) [inline test/const_node_inspector_test.rb:8]: why this change?

**Review by exterm — APPROVED** [review]

File: test/const_node_inspector_test.rb (hunk 6 above)
Comment (mclark) [reply test/const_node_inspector_test.rb:8]: The tests in these two classes started writing formatted output, probably as a result of the Dummy app being initialized. This initially created a log file in the dummy skeleton until I changed it to `stdout`. To silence them completely, I figured I'd just make them consistent with the rest of the tests.
Alternatively, maybe we could configure the Dummy app with some sort of `NullLogger` I suppose?

Comment (exterm) [reply test/const_node_inspector_test.rb:8]: That seems like a good solution 🤔  TBH I don't remember why these tests are using `ActiveSupport::TestCase` and others do not. But it seems like it would be confusing for someone to start using `ActiveSupport::TestCase` and thereby inadvertently generating logs.

Comment (mclark) [reply test/const_node_inspector_test.rb]: Done in 805a77d14179e742ada7ff7a51ad6f19171e7329 👍

**Comment by exterm** [comment]
our main monolith CI is green against this branch. I've also done some manual testing and it's looking great.
